tests/src/sanity.rs

```diff
@@ -0,0 +1,326 @@
+        // 1
+        assert_eq!(jobs.items.len(), 1);
+        let job_name = jobs.items[0].metadata.name.clone().unwrap();
+
+        let pods_api: Api<Pod> = Api::namespaced(client.clone(), "default");
+        let pods = pods_api.list(&ListParams::default()).await.unwrap();
+        let pod_name = pods
+            .items
+            .iter()
+            .filter_map(|pod| {
+                if pod.metadata.name.clone().unwrap().contains(&job_name) {
+                    Some(pod.metadata.name.clone().unwrap())
+                } else {
+                    None
+                }
+            })
+            .next()
+            .unwrap();
+        assert!(pod_name.contains(&job_name));
+        sleep(Duration::from_secs(35)).await;
+
+        //verify cleanup
+        let updated_jobs = jobs_api.list(&ListParams::default()).await.unwrap();
+        assert_eq!(updated_jobs.items.len(), 0);
+
+        let updated_pods = pods_api.list(&ListParams::default()).await.unwrap();
+        assert_eq!(updated_pods.items.len(), 1); // only the nginx pod should exist
+    }
+
+    #[tokio::test]
+    // we send a request to a different pod in the cluster (different namespace) and assert
+    // that no operation is performed as specified in the request by the server
+    // as the agent pod is impersonating the pod running in the default namespace
+    async fn test_different_pod_in_cluster() {
+        let client = setup_kube_client().await;
+
+        let namespace = "test-namespace";
+        create_namespace(&client, namespace).await;
+        create_nginx_pod(&client, namespace).await;
+        sleep(Duration::from_secs(2)).await; // we need a short sleep here, otherwise the test panics
+
+        let service_url = get_service_url(&client, namespace).await.unwrap();
+        let pod_name = get_nginx_pod_name(&client, "default").await.unwrap();
+        let command = vec!["node", "node-e2e/app.js"];
+        let env: HashMap<&str, &str> = HashMap::new();
+        let mut server = start_node_server(&pod_name, command, env);
+
+        let mut stderr_reader = BufReader::new(server.stderr.take().unwrap());
+        setup_panic_hook();
+
+        tokio::spawn(async move {
+            loop {
+                let mut error_stream = String::new();
+                stderr_reader.read_line(&mut error_stream).await.unwrap();
+                if !error_stream.is_empty() {
+                    panic!("Error: {}", error_stream);
+                }
+            }
+        });
+
+        let mut reader = BufReader::new(server.stdout.take().unwrap());
+        let mut line = String::new();
+
+        let timeout_duration = Duration::from_secs(10);
+        timeout(timeout_duration, reader.read_line(&mut line))
+            .await
+            .unwrap()
+            .unwrap();
+
+        assert!(line.contains("Server listening on port 80"));
+        reqwest_request(service_url.as_str(), Method::PUT).await;
```

Comment:
Lets try to keep all the implicit knowledge of what the API does in one place. You already have validate_requests, so maybe add validate_no_requests next to it, or something like that.

Reply:
Just to elaborate - right now, if I make a change to how app.js work, I would have to go over the entire code and look for dependencies. So instead we should strive to cenralize all code that relies on the internal implementation of app.js in one place.